File: app/models/terrain.rb
class Terrain < ActiveHash::Base
    include ActiveHash::Enum
    enum_accessor :name
    self.data = [
        {
            id: 'desert',
            name: 'Desert'
        },
        {
            id: 'fields',
            name: 'Fields',
            resource: Resource::GRAIN
        },
        {
            id: 'pasture',
            name: 'Pasture',
            resource: Resource::WOOL
        },
        {
            id: 'forest',
            name: 'Forest',
            resource: Resource::LUMBER
        },
        {
            id: 'mountains',
            name: 'Mountains',
            resource: Resource::ORE
        },
        {
            id: 'hills',
            name: 'Hills',
            resource: Resource::BRICK
        }
    ]

    def production?
        resource.present?
    end

    def desert?
        self == DESERT
    end
end

File: app/javascript/models/Terrain.ts
export enum TerrainId {
    Desert = 'desert',
    Fields = 'fields',
    Pasture = 'pasture',
    Forest = 'forest',
    Mountains = 'mountains',
    Hills = 'hills'
}

type ProductionTerrainId = TerrainId.Fields | TerrainId.Forest | TerrainId.Hills | TerrainId.Mountains | TerrainId.Pasture

export interface ProductionTerrain {
    id: ProductionTerrainId
}

export interface DesertTerrain {
    id: TerrainId.Desert
}

export type Terrain = ProductionTerrain | DesertTerrain
